lib/Dialect/ModArith/Transforms/ConvertToMac.cpp

```diff
@@ -0,0 +1,58 @@
+#include "lib/Dialect/ModArith/Transforms/ConvertToMac.h"
+
+#include "lib/Dialect/ModArith/IR/ModArithOps.h"
+#include "llvm/include/llvm/Support/Debug.h"            // from @llvm-project
+#include "mlir/include/mlir/IR/ImplicitLocOpBuilder.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+
+#define DEBUG_TYPE "mod-arith-mac"
+
+namespace mlir {
+namespace heir {
+namespace mod_arith {
+
+#define GEN_PASS_DEF_CONVERTTOMAC
+#include "lib/Dialect/ModArith/Transforms/Passes.h.inc"
+
+struct FindMac : public OpRewritePattern<mod_arith::AddOp> {
+  FindMac(mlir::MLIRContext *context)
+      : OpRewritePattern<mod_arith::AddOp>(context) {}
+
+  LogicalResult matchAndRewrite(mod_arith::AddOp op,
+                                PatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    auto *parent = op.getLhs().getDefiningOp();
+
+    if (!parent || !isa<mod_arith::MulOp>(parent)) {
+      return failure();
+    }
+
+    auto mul_parent = cast<mod_arith::MulOp>(parent);
+
+    auto result =
+        b.create<MacOp>(mul_parent.getLhs(), mul_parent.getRhs(), op.getRhs());
+
+    rewriter.replaceOp(op, result);
+    rewriter.eraseOp(mul_parent);
```

Comment:
This is only legal if the parent op has no other uses. Does it still make sense to insert a MAC op if the parent mul result is used in some other location? Then you're just replacing an add op with a MAC op, not strictly reducing op counts.

I think it probably is still useful (in which case do the replacement and conditionally erase), but if not you can alternatively have the match fail if the parent fails `hasSingleUse`.

Reply:
Mhh, the idea of the pass is mostly to find the MAC operations so that they can be lowered more easily. 
For now, I have removed the easeOp, and count that the automatic DCE will remove the multiply operation if it is not required.

Reply:
It's a common pattern to erase if uses is empty. I think that'd be better here than relying on dce being called later